v1/writable/image.go

```diff
@@ -0,0 +1,259 @@
+// Copyright 2018 Google LLC All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package writable
+
+import (
+	"compress/gzip"
+	"encoding/json"
+	"io"
+	"os"
+
+	"github.com/google/go-containerregistry/v1"
+	"github.com/google/go-containerregistry/v1/partial"
+	"github.com/google/go-containerregistry/v1/types"
+	"github.com/google/go-containerregistry/v1/v1util"
+)
+
+type Image struct {
+	v1.Image
+	configFile *v1.ConfigFile
+	manifest   *v1.Manifest
+
+	appendMap map[string]*appendedLayer
+}
+
+type appendedLayer struct {
+	path           string
+	compressed     bool
+	compressedSize int64
+	digest         v1.Hash
+	diffID         v1.Hash
+}
+
+// Digest returns the Hash of the compressed layer.
+func (l *appendedLayer) Digest() (v1.Hash, error) {
+	return l.digest, nil
+}
+
+// DiffID returns the Hash of the uncompressed layer.
+func (l *appendedLayer) DiffID() (v1.Hash, error) {
+	return l.diffID, nil
+}
+
+// Compressed returns an io.ReadCloser for the compressed layer contents.
+func (l *appendedLayer) Compressed() (io.ReadCloser, error) {
+	file, err := os.Open(l.path)
+	if err == nil && !l.compressed {
+		return v1util.GzipReadCloser(file)
+	}
+
+	return file, err
+}
+
+// Uncompressed returns an io.ReadCloser for the uncompressed layer contents.
+func (l *appendedLayer) Uncompressed() (io.ReadCloser, error) {
+	file, err := os.Open(l.path)
+	if err == nil && l.compressed {
+		return v1util.GunzipReadCloser(file)
+	}
+
+	return file, err
+}
+
+// Size returns the compressed size of the Layer.
+func (l *appendedLayer) Size() (int64, error) {
+	return l.compressedSize, nil
+}
+
+func NewImage(base v1.Image) (*Image, error) {
+	m, err := base.Manifest()
+	if err != nil {
+		return nil, err
+	}
+
+	cf, err := base.ConfigFile()
+	if err != nil {
+		return nil, err
+	}
+
+	return &Image{
+		Image:      base,
+		configFile: cf.DeepCopy(),
+		manifest:   m.DeepCopy(),
+		appendMap:  make(map[string]*appendedLayer),
+	}, nil
+}
+
+func (i *Image) AppendLayer(tarFile string) error {
+	var err error
+
+	d := v1.Descriptor{
+		MediaType: types.DockerLayer,
+	}
+
+	if d.Digest, d.Size, err = i.computeDigest(tarFile); err != nil {
+		return err
+	}
+
+	diffID, compressed, err := i.computeDiffID(tarFile)
+	if err != nil {
+		return err
+	}
+
+	// diff id needs to be the uncompressed blob digest
+	i.configFile.RootFS.DiffIDs = append(i.configFile.RootFS.DiffIDs, diffID)
+	i.manifest.Layers = append(i.manifest.Layers, d)
```

Comment:
I believe there are special provisions here for empty layers, which we should handle.

https://github.com/google/containerregistry/blob/master/transform/v2_2/metadata_.py#L206-L212

Reply:
We might also want to update the `history` block.